ntp-daemon/src/peer_manager.rs

```diff
@@ -78,41 +153,45 @@ impl<C: NtpClock> Peers<C> {
         }
     }
 
-    async fn add_peer_internal(&mut self, config: Arc<PeerConfig>) -> JoinHandle<()> {
+    /// Add a single peer to an existing pool
+    async fn add_pool_peer_internal(
+        &mut self,
+        address: NormalizedAddress,
+        mut cached: CachedPoolAddresses,
+    ) -> JoinHandle<()> {
         let index = self.indexer.get();
-        let addr = loop {
-            let host = match &*config {
-                PeerConfig::Standard(StandardPeerConfig { addr, .. }) => {
-                    debug!(unresolved = ?&addr, "lookup host");
-                    addr.lookup_host().await.map(|mut i| i.next())
-                }
-
-                PeerConfig::Pool(PoolPeerConfig { addr, .. }) => {
-                    debug!(unresolved = ?&addr, "lookup host");
-                    addr.lookup_host().await.map(|mut i| i.next())
-                }
-            };
 
-            match host {
-                Ok(Some(addr)) => {
-                    debug!(resolved=?addr, "resolved peer");
-                    break addr;
-                }
-                Ok(None) => {
-                    warn!("Could not resolve peer address, retrying");
-                    tokio::time::sleep(NETWORK_WAIT_PERIOD).await
-                }
-                Err(e) => {
-                    warn!(error = ?e, "error while resolving peer address, retrying");
-                    tokio::time::sleep(NETWORK_WAIT_PERIOD).await
+        // socket addresses of the peers of this pool that are currently active
+        let active_pool_peers = self.peers.values().filter_map(|p| match &p.peer_address {
+            PeerAddress::Peer { .. } => None,
+            PeerAddress::Pool {
+                address: peer_address,
+                socket_address,
+                ..
+            } => {
+                if &address == peer_address {
+                    Some(*socket_address)
+                } else {
+                    None
                 }
             }
+        });
+
+        // if no socket addresses are cached, do a new DNS resolve
+        let addr = match cached.find_additional(&address, active_pool_peers).await {
```

Comment:
I think there is a race condition here where `cached` is not locked, but the `addr` is not yet added to `peers` either. That means that technically another thread could pick the same `addr`, and we duplicate a peer. However, we cannot keep `cached` locked because it is part of the `PeerAddress::Pool` below. Not sure how to deal with that yet.

Reply:
I think this is actually an indication that the state is still too fragmented. Perhaps it is better to reorganize everything such that the list of active pool peers and the cache of unused addresses are part of the same mutex. Access/modification of them is related already so from a performance standpoint that is not that problematic, and it makes correctness like here easier.

Reply:
that is kind of what `attempt-2` did? unless you're implying that pool peers should not be in the list of peers any more?

Reply:
No they definitely should be, I just think that the identification of which of the peers are part of the pool should be part of the pool state.

Reply:
However, now that I look into more detail here, the race condition is a bit unclear to me. The list of peers cant change when this function runs outside of what this function does since we have a self reference.

Reply:
I think that's right, in practice. In theory the `Arc` could have been cloned, but all clones are contained in the `Peers` structure so we should be in the clear.